S0ix KWD PM and dapm even handling

keyonjie opened the pull request:
This is only something verified work with KWD on S0ix, it is very stupid idea but I haven't figured out a decent solution for it yet, so here sending the RFC to try to get better idea.

The problem we need to fix here is that when the system is entering S2idle(e.g. via echo freeze > /sys/power/state), the applications are frozen, and the SNDRV_PCM_TRIGGER_SUSPEND will happen for each PCM substream, which will actually stop all streams in today's code.

But for KWD PCM substream, we need it(or at least the detect pipeline) keep running inside the FW(pipeline 9 in below image), but the detect pipeline state is controlled via DAPM widget event, so here in a bunch, we need to find a way to ignore the DAPM PM events also.

I also thought of another solution, e.g. using .ignore_suspend flag in the corresponding FW dai_link, or making ignore_suspend for those related DAPM widgets. I tried the first one and yes it can work, but haven't tried the latter one yet.

@plbossart @lgirdwood @ranj063 please help give your comments here, I will give some more tries tomorrow before the final solution is chosen.

![tplg](https://user-images.githubusercontent.com/10458135/65474628-27ce8600-deaf-11e9-9e73-647e69885228.png)

plbossart commented:
@keyonjie I thought this was precisely what the .ignore_suspend field in the DMIC input was for?
Why would we need a different definition?

keyonjie commented:
> @keyonjie I thought this was precisely what the .ignore_suspend field in the DMIC input was for?
> Why would we need a different definition?

@plbossart for DMIC input .ignore_suspend field, are you talking about what we are using as "snd_soc_dapm_ignore_suspend(&rtd->card->dapm, "SoC DMIC");" in some machine drivers?

plbossart commented:
On 9/24/19 3:57 PM, Keyon wrote:
>     @keyonjie <https://github.com/keyonjie> I thought this was precisely
>     what the .ignore_suspend field in the DMIC input was for?
>     Why would we need a different definition?
> 
> @plbossart <https://github.com/plbossart> for DMIC input .ignore_suspend 
> field, are you talking about what we are using as 
> "snd_soc_dapm_ignore_suspend(&rtd->card->dapm, "SoC DMIC");" in some 
> machine drivers?

no, I was talking about this:

	{
		.name = "dmic01",
		.id = 2,
		.ignore_suspend = 1, <<<< HERE!
		.be_hw_params_fixup = geminilake_dmic_fixup,
		.dpcm_capture = 1,
		.no_pcm = 1,
		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
	},

keyonjie commented:
> On 9/24/19 3:57 PM, Keyon wrote: @keyonjie <https://github.com/keyonjie> I thought this was precisely what the .ignore_suspend field in the DMIC input was for? Why would we need a different definition? @plbossart <https://github.com/plbossart> for DMIC input .ignore_suspend field, are you talking about what we are using as "snd_soc_dapm_ignore_suspend(&rtd->card->dapm, "SoC DMIC");" in some machine drivers?
> no, I was talking about this: { .name = "dmic01", .id = 2, .ignore_suspend = 1, <<<< HERE! .be_hw_params_fixup = geminilake_dmic_fixup, .dpcm_capture = 1, .no_pcm = 1, SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform), },

Yes, that is for ignoring the invokes to trigger suspend of BE dai_link, let me try the similar one for FE dai_link to see if it work, and, we need similar for DAPM widget also, not sure if .ignore_suspend  in dapm widget can avoid calling the POST_PMU and PRE_PMD event handling.

keyonjie commented:
@plbossart just checked that the .ignore_suspend in struct snd_soc_dapm_widget can't prevent the invoking to the event handler with SND_SOC_DAPM_PRE_PMU and SND_SOC_DAPM_POST_PMD, so I haven't had better solution for this yet.

plbossart commented:
On 9/24/19 6:57 PM, Keyon wrote:
> @plbossart <https://github.com/plbossart> just checked that the 
> .ignore_suspend in struct snd_soc_dapm_widget can't prevent the invoking 
> to the event handler with SND_SOC_DAPM_PRE_PMU and 
> SND_SOC_DAPM_POST_PMD, so I haven't had better solution for this yet.

I honestly have no idea what you are referring to.
It may seem obvious to you but you haven't really shared any clear 
problem statement so far...

keyonjie commented:
> I honestly have no idea what you are referring to. It may seem obvious to you but you haven't really shared any clear problem statement so far...

@plbossart it's here: https://github.com/thesofproject/linux/blob/e03f8829f95c175f63c808f5747b977e7c443765/sound/soc/sof/topology.c#L134

When we run 'echo freeze > /sys/power/state', this DAPM event handler will be called and then the KWD detect pipeline will be stopped, via https://github.com/thesofproject/linux/blob/e03f8829f95c175f63c808f5747b977e7c443765/sound/soc/sof/topology.c#L170

But that's not what we want, we want to keep the detect pipeline(for key phrase detect) running, that means we should not send trigger stop and pcm free IPCs to FW.

Above is the problem we hit, and the PR here is aimed to fix it.

plbossart commented:
thanks @keyonjie the pointers were very helpful to understand the problem.
Couple of comments:
0. you cannot make any references to an ACPI target state in the SOF core. We need to have a solution that compiles/works for non-ACPI systems.
1. I have a bit of heartburn with the 'global' way of handling events with the field handled at the sdev level. can't we have some sort of private data or link between widget and PCM stream? I see in the topology that we connect the DAPM widget with the PCM stream, so can't we check if there's a connection with a D0i3 compatible PCM, and make the ignore_dapm event more 'local' and tied to the PCM device?

keyonjie commented:
> thanks @keyonjie the pointers were very helpful to understand the problem.
> Couple of comments:
> 0. you cannot make any references to an ACPI target state in the SOF core. We need to have a solution that compiles/works for non-ACPI systems.

That's a good point, let me think about how to handle it.

> 
> 1. I have a bit of heartburn with the 'global' way of handling events with the field handled at the sdev level. can't we have some sort of private data or link between widget and PCM stream? I see in the topology that we connect the DAPM widget with the PCM stream, so can't we check if there's a connection with a D0i3 compatible PCM, and make the ignore_dapm event more 'local' and tied to the PCM device?

Yes, that's my feel also, let me try to figure out to remove this from global context.

keyonjie commented:
@plbossart updated.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -307,6 +307,7 @@ struct snd_sof_pcm_stream {
 	struct snd_pcm_substream *substream;
 	struct work_struct period_elapsed_work;
 	bool d0i3_compatible; /* DSP can be in D0I3 when this pcm is opened */
+	bool d0i3_active; /* This stream is active at D0I3 */
```
you are not setting this flag in this PR, so it's hard to figure out what it really represents.

You are really confusing concepts and terminology

a device is active/suspended
a stream is opened/prepared/triggered.

an active stream is ambiguous.

keyonjie replied:
I think in ALSA concept, we also have 'active' concept for a PCM stream, where when the stream is in opened or prepared or running, haven't we? I didn't use 'd0i3_opened' as it sounds confusing to me, or what wording do you suggest here?

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -363,6 +366,12 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_START;
 		break;
 	case SNDRV_PCM_TRIGGER_SUSPEND:
+		/* Don't suspend the d0i3 compatible stream */
```
well it depends on the target state. If you target S0, then you should not suspend. If you target S3, then yes you should absolutely suspend.

When I made a reference in earlier reviews that we cannot have explicit acpi references, it doesn't mean we need to forget about the concept. I was really asking for an abstraction that compiles when CONFIG_ACPI is not selected.

keyonjie replied:
I got your point, I intended to make logic simpler with the assumption that when d0i3 stream opened, we will only enter d0i3 and never do power off to the ADSP. The cons of this is that we can't handle S3 suspend for audio part.

Okay, let me keep the check with fixing the compiling issue when CONFIG_ACPI not selected.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -146,11 +148,20 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	/* process events */
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
+		if (spcm->stream[stream].d0i3_active) {
```
why would we test only for capture?

What if capture does not exist?

keyonjie replied:
Here the function sof_keyword_dapm_event is for KWD which is capture only, it existed in today's code like that 
```
line 153: ret = ipc_pcm_params(swidget, SOF_IPC_STREAM_CAPTURE);
```
So, we implement this event handler with assumption that there is and only is capture stream.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -146,11 +148,20 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
+		if (spcm->stream[stream].d0i3_active) {
+			dev_dbg(sdev->dev, "KWD worked in suspended, no need to start it\n");
```
Remove references to KWD, this is a generic capability that can be used for other usages.

keyonjie replied:
We don't need this DAPM event handler of other use case e.g. deep-buffer playback, do we?

kv2019i replied:
@keyonjie @plbossart  This is a KWD event handled, so the KWD prefix seems appropriate. I'd change the wording though to something like: "KWD stream active in D0ix, no need to resume" ... "worked in suspended" is a bit confusing

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -146,11 +148,20 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	case SND_SOC_DAPM_PRE_PMU:
+		if (spcm->stream[stream].d0i3_active) {
+			dev_dbg(sdev->dev, "KWD worked in suspended, no need to start it\n");
+			spcm->stream[stream].d0i3_active = false;
```
why would you turn it off here?

keyonjie replied:
Maybe I should add explain about how this work in the commit message:

1. PCM trigger suspend (set d0i3_active) -> 
2. DAPM event handler for POST_PMD (d0i3_active set, ignore the event) --> 
3. DAPM event handler for PRE_PMU (d0i3_active set, no need to handle this power up event) -> 
4. PCM trigger resume (don't do real resuming as d0i3_active already set, clear d0i3_active after done).

I originally put this turning off in step 4 as the steps stated above, but then I realize this PCM trigger resume will never happen with our driver, that's why move it to the earlier step 3. 

Hi @ranj063 do you know why pcm trigger resume will never happen with our driver, I recall you did something related (suspend_on_resume?)?

kv2019i replied:
@keyonjie The logic seems fine. Setting/clearing d0i3_active seems a bit complicated, but I understand we need some mechanism to distribute this information to widget events. I'd be also ok with "d0i3_ignore_dapm_events" that is set unless we are going to S3.

ranj063 commented:
@keyonjie ignore_suspend should prevent the DAPM event PRE_PMU and POST_PMD. have you tried it already?

keyonjie commented:
> @keyonjie ignore_suspend should prevent the DAPM event PRE_PMU and POST_PMD. have you tried it already?

I tried, it doesn't. I think the .ignore_suspend of snd_soc_dapm_widget is only helpful at keeping the widget itself enabled over suspend, but there is no relation with those DAPM events.

Hi @lgirdwood is that true?

ranj063 commented:
> I tried, it doesn't. I think the .ignore_suspend of snd_soc_dapm_widget is only helpful at keeping the widget itself enabled over suspend, but there is no relation with those DAPM events.

@keyonjie I meant ignore_suspend for the dai_link not the widget itself.

keyonjie commented:
> > I tried, it doesn't. I think the .ignore_suspend of snd_soc_dapm_widget is only helpful at keeping the widget itself enabled over suspend, but there is no relation with those DAPM events.
> 
> @keyonjie I meant ignore_suspend for the dai_link not the widget itself.

ah, I also tried that, that will help for not sending SNDRV_PCM_TRIGGER_SUSPEND in pcm_trigger only, but it can't prevent all the DAPM POST_PMD and PRE_PMU events, check it here:

	for_each_card_rtds(card, rtd) {

		if (rtd->dai_link->ignore_suspend)
			continue;

		snd_soc_dapm_stream_event(rtd,
					  SNDRV_PCM_STREAM_PLAYBACK,
					  SND_SOC_DAPM_STREAM_SUSPEND);

		snd_soc_dapm_stream_event(rtd,
					  SNDRV_PCM_STREAM_CAPTURE,
					  SND_SOC_DAPM_STREAM_SUSPEND);//Keyon: here the stream event won't be sent if ignore_suspend is specified, but...
	}

	/* Recheck all endpoints too, their state is affected by suspend */
	dapm_mark_endpoints_dirty(card);
	snd_soc_dapm_sync(&card->dapm);//Keyon: another round of events will be sent here, triggered with SND_SOC_DAPM_STREAM_NOP event.

ranj063 commented:
@keyonjie sorry took me a while to process this PR but the logic here makes sense. The confusion arises from the name of d0i3_active. If you call it something like ignore_dapm_event instead, then it would clarify what the intention is. So when the system suspends, you would prevent handling the DAPM events for the widgets associated with d0i3_compatible streams.

keyonjie commented:
> @keyonjie sorry took me a while to process this PR but the logic here makes sense. The confusion arises from the name of d0i3_active. If you call it something like ignore_dapm_event instead, then it 

I am good with 'ignore_dapm_event', it is actually the exact name in my original PR in 30th August, where the flag is added to sof-dev:

```
diff --git a/sound/soc/sof/sof-priv.h b/sound/soc/sof/sof-priv.h
index e8ea492fdd75..25ed0d7032f7 100644
--- a/sound/soc/sof/sof-priv.h
+++ b/sound/soc/sof/sof-priv.h
@@ -39,6 +39,9 @@
 /* time in ms for runtime suspend delay */
 #define SND_SOF_SUSPEND_DELAY_MS	2000
 
+/* delay before scheduling D0i3 entry work */
+#define SND_SOF_D0I3_DELAY 5000
+
 /* DMA buffer size for trace */
 #define DMA_BUF_SIZE_FOR_TRACE (PAGE_SIZE * 16)
 
@@ -384,7 +387,8 @@ struct snd_sof_dev {
 	enum sof_dsp_state dsp_state;
 	u32 pcm_active;		/* Count of active PCMs */
 	u32 lp_pcm_active;	/* Count of active low power PCMs */
-	struct work_struct d0ix_work;
+	struct delayed_work d0i3_entry_work;
+	bool ignore_dapm_event;
 
 	/* DSP firmware boot */
 	wait_queue_head_t boot_wait;
```

> would clarify what the intention is. So when the system suspends, you would prevent handling the DAPM events for the widgets associated with d0i3_compatible streams.

keyonjie commented:
Hi Pierre and Ranjani,

As we have talked in the call, I tracked the suspend/resume sequence, to summarize, the sequence happens like this:
```
1. run arecord for KWD stream:
DAPM PRE_PMU,
DAPM POST_PMU,
sof-pcm trigger start

2. echo freeze > /sys/power/state:
snd_soc_suspend ->
     sof-pcm trigger suspend,
     DAPM PRE PMD,
     DAPM POST PMD,
snd_sof_suspend   //we check acpi target state here

3. press power key or tap the dmic to resume:
snd_sof_resume,
snd_soc_resume ->
     DAPM PRE PMU,
     DAPM POST PMU,
     sof-pcm trigger start.   // no trigger resume happen.

```

----dmesg log start----
```
root@test-Hatch:~# dmesg -w | grep Keyon
[   35.752400] Keyon: sof_keyword_dapm_event, 148, received event 1 for widget DETECT9.0(type:33)
[   35.752616] Keyon: sof_keyword_dapm_event, 148, received event 2 for widget DETECT9.0(type:33)
[   35.752883] Keyon: sof_pcm_trigger, 337, pcm: trigger stream 8 dir 1 cmd 1
[   46.035383] Keyon: snd_soc_suspend, 443
[   46.035404] Keyon: sof_pcm_trigger, 337, pcm: trigger stream 8 dir 1 cmd 5
[   46.035656] Keyon: sof_keyword_dapm_event, 148, received event 4 for widget DETECT9.0(type:33)
[   46.035665] Keyon: sof_keyword_dapm_event, 148, received event 8 for widget DETECT9.0(type:33)
[   46.035971] Keyon: snd_sof_suspend, 477, acpi target state:0
[   54.654793] Keyon: snd_sof_resume, 442, acpi target state:0
[   54.655012] Keyon: snd_soc_resume, 675
[   54.667433] Keyon: sof_keyword_dapm_event, 148, received event 1 for widget DETECT9.0(type:33)
[   54.667445] Keyon: sof_keyword_dapm_event, 148, received event 2 for widget DETECT9.0(type:33)
[   54.772490] Keyon: sof_pcm_trigger, 337, pcm: trigger stream 8 dir 1 cmd 1
```

---debug patch start----
```
Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>
---
 sound/Makefile           | 2 +-
 sound/soc/soc-core.c     | 2 ++
 sound/soc/sof/ipc.c      | 8 ++++----
 sound/soc/sof/pcm.c      | 2 ++
 sound/soc/sof/pm.c       | 1 +
 sound/soc/sof/topology.c | 7 +++++--
 6 files changed, 15 insertions(+), 7 deletions(-)

diff --git a/sound/Makefile b/sound/Makefile
index 797ecdcd35e2..c3b05c8687eb 100644
--- a/sound/Makefile
+++ b/sound/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 # Makefile for the Linux sound card driver
 #
-
+export KBUILD_CFLAGS += -DDEBUG
 obj-$(CONFIG_SOUND) += soundcore.o
 obj-$(CONFIG_DMASOUND) += oss/dmasound/
 obj-$(CONFIG_SND) += core/ i2c/ drivers/ isa/ pci/ ppc/ arm/ sh/ synth/ usb/ \
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index 35f48e9c5ead..53a236c0462b 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -440,6 +440,7 @@ int snd_soc_suspend(struct device *dev)
     struct snd_soc_pcm_runtime *rtd;
     int i;

+    printk(KERN_DEBUG "Keyon: %s, %d\n", __func__, __LINE__);
     /* If the card is not initialized yet there is nothing to do */
     if (!card->instantiated)
         return 0;
@@ -671,6 +672,7 @@ int snd_soc_resume(struct device *dev)
     struct snd_soc_dai *codec_dai;
     int i;

+    printk(KERN_DEBUG "Keyon: %s, %d\n", __func__, __LINE__);
     /* If the card is not initialized yet there is nothing to do */
     if (!card->instantiated)
         return 0;
diff --git a/sound/soc/sof/ipc.c b/sound/soc/sof/ipc.c
index 26675dfe0240..9c8c732bc6ce 100644
--- a/sound/soc/sof/ipc.c
+++ b/sound/soc/sof/ipc.c
@@ -334,7 +334,7 @@ void snd_sof_ipc_msgs_rx(struct snd_sof_dev *sdev)

     /* read back header */
     snd_sof_ipc_msg_data(sdev, NULL, &hdr, sizeof(hdr));
-    ipc_log_header(sdev->dev, "ipc rx", hdr.cmd);
+//    ipc_log_header(sdev->dev, "ipc rx", hdr.cmd);

     cmd = hdr.cmd & SOF_GLB_TYPE_MASK;
     type = hdr.cmd & SOF_CMD_TYPE_MASK;
@@ -381,7 +381,7 @@ void snd_sof_ipc_msgs_rx(struct snd_sof_dev *sdev)
         break;
     }

-    ipc_log_header(sdev->dev, "ipc rx done", hdr.cmd);
+//    ipc_log_header(sdev->dev, "ipc rx done", hdr.cmd);
 }
 EXPORT_SYMBOL(snd_sof_ipc_msgs_rx);

@@ -428,8 +428,8 @@ static void ipc_period_elapsed(struct snd_sof_dev *sdev, u32 msg_id)
     stream = &spcm->stream[direction];
     snd_sof_ipc_msg_data(sdev, stream->substream, &posn, sizeof(posn));

-    dev_dbg(sdev->dev, "posn : host 0x%llx dai 0x%llx wall 0x%llx\n",
-        posn.host_posn, posn.dai_posn, posn.wallclock);
+//    dev_dbg(sdev->dev, "posn : host 0x%llx dai 0x%llx wall 0x%llx\n",
+//        posn.host_posn, posn.dai_posn, posn.wallclock);

     memcpy(&stream->posn, &posn, sizeof(posn));

diff --git a/sound/soc/sof/pcm.c b/sound/soc/sof/pcm.c
index ff735fb16f03..0748430574a2 100644
--- a/sound/soc/sof/pcm.c
+++ b/sound/soc/sof/pcm.c
@@ -334,6 +334,8 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
     if (!spcm)
         return -EINVAL;

+    printk(KERN_DEBUG "Keyon: %s, %d, pcm: trigger stream %d dir %d cmd %d\n", __func__, __LINE__, spcm->pcm.pcm_id, substream->stream, cmd);
+
     dev_dbg(sdev->dev, "pcm: trigger stream %d dir %d cmd %d\n",
         spcm->pcm.pcm_id, substream->stream, cmd);

diff --git a/sound/soc/sof/pm.c b/sound/soc/sof/pm.c
index 0426e3f55bde..144263a7fe84 100644
--- a/sound/soc/sof/pm.c
+++ b/sound/soc/sof/pm.c
@@ -439,6 +439,7 @@ int snd_sof_resume(struct device *dev)
     struct pci_dev *pci = to_pci_dev(sdev->dev);
     int ret;

+    printk(KERN_DEBUG "Keyon: %s, %d, acpi target state:%d\n", __func__, __LINE__, acpi_target_system_state());
     if (sdev->d0_substate == SOF_DSP_D0I3) {
         dev_dbg(sdev->dev, "ADSP will exit from D0i3...\n");
         ret = snd_sof_set_d0_state(sdev, SOF_DSP_D0I0);
diff --git a/sound/soc/sof/topology.c b/sound/soc/sof/topology.c
index 15a43f7d0bd8..a3afecd2e02f 100644
--- a/sound/soc/sof/topology.c
+++ b/sound/soc/sof/topology.c
@@ -145,8 +145,10 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,

     sdev = swidget->sdev;

-    dev_dbg(sdev->dev, "received event %d for widget %s\n",
-        event, w->name);
+    printk(KERN_DEBUG "Keyon: %s, %d, received event %d for widget %s(type:%d)\n", __func__, __LINE__, event, w->name, w->id);
+
+    dev_dbg(sdev->dev, "received event %d for widget %s(type:%d)\n",
+        event, w->name, w->id);

     /* get runtime PCM params using widget's stream name */
     spcm = snd_sof_find_spcm_name(sdev, swidget->widget->sname);
@@ -2868,6 +2870,7 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
      */
     if (!link->no_pcm) {
         link->nonatomic = true;
+//        link->ignore_suspend = 1;

         /* nothing more to do for FE dai links */
         return 0;
-- 
2.20.1
```

keyonjie commented:
I also spent hours to try to move binding of the DAPM event to our virtual widget, but failed (it require bit change in both topology and driver, e.g. we need make sroute widget->widget legal, change the route load, implement helper to find the detect component from the virtual component, change the runtime resume to not send connection of virtual route to FW, ...).

keyonjie commented:
So finally I gave up and prefer to use the solution in my PR #1249.

please let me know what do you think about it.

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -166,6 +177,11 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 		break;
 	case SND_SOC_DAPM_POST_PMD:
+		if (spcm->stream[stream].d0i3_active) {
+			dev_dbg(sdev->dev, " KWD will work in suspended, don't suspend it\n");
```
Minor nit: "KWD stream will be active in D0ix, not suspending"

keyonjie replied:
Thanks @kv2019i , but @plbossart mentioned several times that wording "active" could be confusing also, I am not a native speaker, but I like your wording here, let me consider do some changes there.
And, I will change the flag to ignore_dapm_event at the same time.

ranj063 commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -150,14 +151,50 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
-	snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
-			  CNL_DSP_REG_HIPCIDR_BUSY);
+	u32 dr = 0;
+	u32 dd = 0;
```
@keyonjie do you need the initialization for these?

keyonjie replied:
yes, otherwise, there will be warning complaining about being possible to use them uninitialized below, as cnl_compact_ipc_assemble() might not touch them sometimes.

ranj063 commented on sound/soc/sof/pm.c:
```diff
@@ -435,12 +435,61 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	if (acpi_target_system_state() == ACPI_STATE_S0) {
```
@keyonjie why check again? cant we use the saved state from prepare?

keyonjie replied:
Ah I actually changed that, don't know why it roll back when squashing, let me fix it now, thanks.

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -363,6 +369,15 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
+		 * Don't suspend the d0i3 compatible stream when target
+		 * is S0ix, and ignore the coming round of DAPM events.
+		 */
+		if (sstream->d0i3_compatible && !sdev->target_state) {
```
better to be explicit here about target_state > ACPI_STATE_S0

keyonjie replied:
OK, that will be sdev->target_state == ACPI_STATE_S0.

ranj063 commented:
@keyonjie this is looking better but I have a question. When does the complete callback happen? before or after PRE_PMU event for DAPM widgets?

If it happens after, why not just set/clear the ignore_dapm_event flag in the prepare/complete callbacks? That way we can avoid saving the target_state

keyonjie commented:
> @keyonjie this is looking better but I have a question. When does the complete callback happen? before or after PRE_PMU event for DAPM widgets?

after.

> 
> If it happens after, why not just set/clear the ignore_dapm_event flag in the prepare/complete callbacks? That way we can avoid saving the target_state

The  ignore_dapm_event belong to each stream here, but yes, let me consider moving it to sdev.

keyonjie commented:
@ranj063 now it is simplified.

ranj063 commented on sound/soc/sof/pm.c:
```diff
@@ -435,12 +435,63 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
+	int ret;
+
+	if (sdev->s0_suspend) {
+		dev_dbg(sdev->dev, "ADSP will try to enter D0i3 for S2idle...\n");
```
@keyonjie this comment can be removed but if you want to keep it, it should be enough to say DSP trying to enter D0i3.

keyonjie replied:
updated.

ranj063 commented:
@keyonjie the logic looks correct but it is a bit hard to review because of the levels of abstraction to actually set the DSP power state.

sof_suspend-> snd_sof_set_S0_state->snd_sof_dsp_set_power_state

Can we remove snd_sof_set_s0_state maybe? All that you do in that function is set the current D0_substate. That can be set in the sof_suspend() call itself. What do you think?

keyonjie commented:
> @keyonjie the logic looks correct but it is a bit hard to review because of the levels of abstraction to actually set the DSP power state.
> 
> sof_suspend-> snd_sof_set_S0_state->snd_sof_dsp_set_power_state
> 
> Can we remove snd_sof_set_s0_state maybe? All that you do in that function is set the current 

I added this snd_sof_set_s0_state with the aim to do general(non-platform specific) stuff and the snd_sof_dsp_set_power_state for platform-specific stuff. Now what we do is the only D0_substate setting, but as you see in the comment message, we will need to do unused pipelines/widgets unregistering to FW in a short future. 

> D0_substate. That can be set in the sof_suspend() call itself. What do you think?

ranj063 commented:
> I added this snd_sof_set_s0_state with the aim to do general(non-platform specific) stuff and the snd_sof_dsp_set_power_state for platform-specific stuff. Now what we do is the only D0_substate setting, but as you see in the comment message, we will need to do unused pipelines/widgets unregistering to FW in a short future.
> 
> >

@keyonjie fair enough. But I think taking down pipelines is something we should do even otherwise in the case of S3 not just for S0ix. You could argue why since we're going to power off the DSP anyway but with the changes that I am making in my MFD PR and looking at splitting up the topology based on the type of DAI, it will be essential.

keyonjie commented:
> > I added this snd_sof_set_s0_state with the aim to do general(non-platform specific) stuff and the snd_sof_dsp_set_power_state for platform-specific stuff. Now what we do is the only D0_substate setting, but as you see in the comment message, we will need to do unused pipelines/widgets unregistering to FW in a short future.
> > >
> 
> @keyonjie fair enough. But I think taking down pipelines is something we should do even otherwise in the case of S3 not just for S0ix. You could argue why since we're going to power off the

That makes sense, I think. Doing the pipelines taking down in the opposite(or symmetric) sequence before shutting down ADSP is what I have been considering, what I had in my mind about that is, implement pipeline taking down helpers, call it at anytime it should be called, e.g. at Doi3 runtime(when the pipeline is not used for a specific time, can be marco defined choice), at D3, before shutting down ADSP.

> DSP anyway but with the changes that I am making in my MFD PR and looking at splitting up the topology based on the type of DAI, it will be essential.

keyonjie commented:
We were blocked in verifying this as the firmware doesn't work as expected in S0ix, will get back once it is verified.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -411,6 +411,28 @@ int snd_sof_runtime_resume(struct device *dev)
 }
 EXPORT_SYMBOL(snd_sof_runtime_resume);
 
+int snd_sof_set_d0_state(struct snd_sof_dev *sdev,
```
should be set_d0_substate if you want to be consistent with the parameters

keyonjie replied:
OK, let me change it.

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -15,6 +15,7 @@
  * Hardware interface for audio DSP on Cannonlake.
  */
 
+#include <sound/sof/header-intel-cavs.h>
```
not added yet, this breaks git bisect.

keyonjie replied:
wow, the github make the commits order totally wrong.

keyonjie replied:
it is correct in my personal branch, maybe I need to create new PR to fix it.

ranj063 replied:
@keyonjie lets fix the commit title to "ASoC: SOF: Intel: CNL: add support for sending compact IPC"

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -150,14 +151,50 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
+		pm_gate = msg->msg_data;
+		*dr = CAVS_IPC_MSG_TGT | CAVS_IPC_MOD_SET_D0IX;
+		*dd = pm_gate->flags & CAVS_IPC_MOD_SETD0IX_BIT_MASK;
```
comments, anyone? What is this supposed to do?

keyonjie replied:
here it set the dd register (IPC extended data register) with the configured flags value. Let me try add some comments.

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -150,14 +151,50 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
+	u32 dr = 0;
+	u32 dd = 0;
+
+	/* send the message via IPC registers only */
```
comment should be below the if()

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -435,12 +435,66 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	/* suspend to D0i3 */
```
comment should be below the if()

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -435,12 +435,66 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
+	int ret;
+
+	/* resuming from D0I3 */
+	if (sdev->d0_substate == SOF_DSP_D0I3) {
```
it's somewhat odd that you are not using the same state variable for suspend and resume.

keyonjie replied:
I wanted to use this only at the case that it is on D0I3, but yes, using s0_suspend here should also work, as the PM framework should already able to handle error case for us, let me change it.

ranj063 replied:
@keyonjie lets use s0_suspend() to keep the code balanced

ranj063 replied:
Also lets fix the title "ASoC: SOF: PM: Add support for DSP D0i3 state when entering S0ix"

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -435,12 +435,66 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
+	int ret;
+
+	/* suspend to D0i3 */
+	if (sdev->s0_suspend) {
```
where is this set? Can there be a race between the time this s0_suspend is set and the time when it's used?

keyonjie replied:
there is no race, it is set in .prepare and .complete, and used between them only.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -435,12 +435,66 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
+		return 0;
+	}
+
+	/* else, deep suspend */
```
'deep' is not defined.
"regular suspend"?

keyonjie replied:
ok, the naming "deep" actually from mem_sleep.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -346,6 +346,11 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
 		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_RELEASE;
 		break;
 	case SNDRV_PCM_TRIGGER_RESUME:
+		/* No need to resume it as the stream is already active */
```
what does active mean?
You keep using this word and you are not listening to feedback.

keyonjie replied:
um, I keep using this as I can't find a suitable one. :( let me change it to "opened"?

ranj063 replied:
@keyonjie just say "No need to resume streams that werent suspended"

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -356,9 +361,21 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
 
 		/* fallthrough */
 	case SNDRV_PCM_TRIGGER_START:
+		/* No need to start it as the stream is already active */
```
is this legit? 
This state machine seems before convoluted since we don't support the INFO_RESUME flag.
So when would we actually get a TRIGGER_START while in D0i3?
That sounds like copy/paste and not really thought through.

keyonjie replied:
> is this legit?
> This state machine seems before convoluted since we don't support the INFO_RESUME flag.

Yes, exactly, as we don't support the INFO_RESUME flag, TRIGGER_RESUME actually won't happen, and replace, TRIGGER_START is sent at resume, that's why I need add check to the TRIGGER_START case also.

> So when would we actually get a TRIGGER_START while in D0i3?
> That sounds like copy/paste and not really thought through.

The trigger_start will be sent when resuming from S0ix(audio in D0i3), so this one here is needed and the trigger resume one can be actually removed(I keep both as once we change to support INFO_RESUME flag, it can still work).

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -146,11 +148,19 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	/* process events */
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
+		if (spcm->stream[stream].suspend_ignored) {
```
the suspend_ignored flag is set based on PCM triggers.
you test it based on DAPM triggers.
How do I know it's consistent and that there's no race condition? Does this even work?

ranj063 replied:
@plbossart let me try to explain the flow in the order in which they happen. @keyonjie correct me if I am wrong:
During suspend: 
1. system suspend prepare callback -> sets S0_suspend flags
2. Streams are suspended with a TRIGGER_SUSPEND -> sets the suspend_ignored flag
3. DAPM widgets are powered down -> POST_PMD event -> if suspend_ignored flag set-> do nothing. else stop the pipeline
4. SOF device suspend callback 

During system resume
1. SOF device resume callback
2. system resume complete callback -> sets the s0_suspend flag to false.
3. DAPM widgets powered up -> check if suspend_ignored and do nothing, else start the pipeline
4. TRIGGER_START for all pipelines -> unset the suspend_ignored flag

Does this make sense?

plbossart replied:
yes, but we don't have a case with TRIGGER_RESUME, so it should be marked with some sort of warning saying the code can't really work.

keyonjie replied:
@plbossart do you mean adding some comments in the trigger_resume case, e.g.:
```
       case SNDRV_PCM_TRIGGER_RESUME:
		/*
                 * Only happens when INFO_RESUME is supported, 
                 * no need to resume streams that were not suspended.
                 */
...
		/* fallthrough */
	case SNDRV_PCM_TRIGGER_START:
		/*
                 * This may arrive at stream resuming where INFO_RESUME is
                 * not supported, no need to re-start it if the stream was
                 * not suspended in this case.
                 */
...
```

plbossart replied:
yes, maybe reworded a bit

````
case SNDRV_PCM_TRIGGER_RESUME:
		/*
                 * this case will be triggered when INFO_RESUME is supported, 
                 * no need to resume streams that remained enabled in D0ix.
                 */
...
		/* fallthrough */
	case SNDRV_PCM_TRIGGER_START:
		/*
                 * This case will be triggered when INFO_RESUME is not supported
                 * no need to re-start streams that remained enabled in D0ix.
                 */

````

keyonjie replied:
> yes, maybe reworded a bit
> 
> ```
> case SNDRV_PCM_TRIGGER_RESUME:
> 		/*
>                  * this case will be triggered when INFO_RESUME is supported, 
>                  * no need to resume streams that remained enabled in D0ix.
>                  */
> ...
> 		/* fallthrough */
> 	case SNDRV_PCM_TRIGGER_START:
> 		/*
>                  * This case will be triggered when INFO_RESUME is not supported
>                  * no need to re-start streams that remained enabled in D0ix.
>                  */
> ```

Wow, the magic of wording, it looks much better and symmetric with this, thank you Pierre.

plbossart commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -336,7 +354,17 @@ int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
-	return 0;
+	/* prevent power gating at D0 */
+	if (d0_substate == SOF_DSP_D0I0)
+		flags = SOF_PM_PPG;
```
don't you have to set flags for D0i3?
this PR seems completely out of order, I am not sure what you did with github...

keyonjie replied:
No, we had explained this, setting 0s for D0i3, that means power gating is permitted(not prevented), makes sense?

plbossart commented on include/sound/sof/header-intel-cavs.h:
```diff
@@ -0,0 +1,82 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
Dual license please. this is GPLv2 OR BSD-3-Clause

keyonjie replied:
OK, thanks.

plbossart commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -20,6 +20,17 @@
+#define SOF_PM_PPG		BIT(2)
+
+/* Prevent clock gating (0 - cg allowed, 1 - DSP clock always on) */
+#define SOF_PM_PCG		BIT(3)
```
so why do we have definitions that seem to duplicate the Intel ones? gah this is so confusing.

 e.g. we had this:

#define CAVS_IPC_MOD_SETD0IX_PCG	BIT(3)

/* Prevent power gating (0 - D0ix transitions allowed) */
#define CAVS_IPC_MOD_SETD0IX_PPG	BIT(2)

keyonjie replied:
I think it is because one is for SOF IPC while the other one is of CAVS IPC, @mmaka1?

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -444,3 +444,27 @@ int snd_sof_suspend(struct device *dev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+
+#if defined(CONFIG_ACPI) && defined(CONFIG_SND_SOC_SOF_DEBUG_D0I3_AT_S0IX)
```
This should be based on a kernel module param with the default value set with a Kconfig.
Also not sure why it's called DEBUG_D0I3, it's really enable D0ix.

keyonjie replied:
The intention here is, 
1. some people might want the audio DSP to enter to D3(at S0ix) even the platform can support D0i3, so the the debug item is "D0i3 at S0ix".
2. eventually, I want to change the default of this item to be 'true', which means turn on "D0i3 at S0ix" if the platform supported. But to make it compatible at this moment for other platforms (to avoid possible massive regressions on some other platforms which are not tested yet).

So this is a flag for the transition period only, finally, it could be a true debug option "D3 at S0ix" which is selected for debug purpose only.

ranj063 replied:
@keyonjie I feel like the DEBUG_D0i3 is a bit of an overkill at this point. Lets leave this for later if really needed. For now, if the DSP supports D0i3, we should enter D0i3.

keyonjie replied:
OK, let me remove that, but we need to keep in mind that the suspend/resume behaviors will be changed for many other platforms, and new issues might be reported.

plbossart commented on sound/soc/sof/sof-pci-dev.c:
```diff
@@ -246,6 +246,8 @@ static const struct sof_dev_desc ehl_desc = {
 
 static const struct dev_pm_ops sof_pci_pm = {
+	.prepare = snd_sof_prepare,
+	.complete = snd_sof_complete,
```
this really feels wrong. we should not add those callbacks for all platforms, but add them as needed.
e.g. this should not be enabled for Merrifield!

keyonjie replied:
What these tow ops do is only setting and clearing s0_suspend flag, to reflect the real target system state, on Merrifield, as we don't support D0i3 there, we can just leave the flag and not use it?

ranj063 replied:
@keyonjie fix commit message and title: "Add prepare/complete PM callbacks for platform X, Y and Z" along with Pierre's suggestions.

And commit message should be similar too. No need to say specific stuff happening....

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -308,6 +308,7 @@ struct snd_sof_pcm_stream {
 	struct snd_pcm_substream *substream;
 	struct work_struct period_elapsed_work;
 	bool d0i3_compatible; /* DSP can be in D0I3 when this pcm is opened */
+	bool suspend_ignored;
```
remove the Change-Id in the commit message

keyonjie replied:
oops, thanks, looks that came from the repo management.

ranj063 replied:
@keyonjie this patch needs a bit more explanation even thoughthe change is really a one-liner.
The commit message should say this so people dont wonder why we have "S0_suspend" and suspend_ignored. 
Also, lets not say stream is active. From a core point of view, the stream is suspended. But, internally the SOF DSP keeps the pipeline associated with it active. There is a big difference between the two. 

Anyway back to my point: This flag is associated with each stream and every stream as it is  suspended, sets the flag to indicate that the pipeline should be kept active and is set based on the system state being entered (which is indicated by the s0_suspend flag).

Could  you please add this to the commit message and the comment above?

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -166,6 +166,16 @@ config SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT
+	help
+	  This enable audio device D0I3 support, where audio can keep partially
+	  active with low power consumption, which is typically used at Wake on
+	  Voice or deep buffer playback scenarios.
```
well you didn't really enabled D0ix, you only transition to D0ix on S0ix transition, so you've really done half of the work.

keyonjie replied:
Yes, we still have runtime D0i3 support task in the to-do list, let me change it to "This enable audio device D0Ix support on S0ix", fine?

ranj063 replied:
@keyonjie I would drop this patch for now.

keyonjie replied:
OK, let me drop this, and make audio entering D0i3 on S0ix certainly when the platform supported(that is, "sof_ops(sdev)->set_power_state != NULL").

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -166,6 +166,16 @@ config SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT
 
+config SND_SOC_SOF_DEBUG_D0I3_AT_S0IX
+	bool "SOF put audio device to D0I3 state at S0Ix"
+	default n
```
remove, all options are default n

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -166,6 +166,16 @@ config SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT
 	  If unsure, select "N".
 
+config SND_SOC_SOF_DEBUG_D0I3_AT_S0IX
+	bool "SOF put audio device to D0I3 state at S0Ix"
```
this is not a debug option, it's a feature.

keyonjie replied:
As commented above, I intended to make it a debug option actually.

plbossart commented:
>> @@ -346,6 +346,11 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
>   		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_RELEASE;
>   		break;
>   	case SNDRV_PCM_TRIGGER_RESUME:
> +		/* No need to resume it as the stream is already active */
> 
> @keyonjie <https://github.com/keyonjie> just say "No need to resume 
> streams that werent suspended"

My point was that since we don't support INFO_RESUME can this really 
happen in the first place?

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -146,11 +148,19 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
+		if (spcm->stream[stream].suspend_ignored) {
+			dev_dbg(sdev->dev, "KWD stream already active in S0ix, no need to resume\n");
```
@keyonjie lets fix all comments to not say active. Lets use suspended or not suspended?

keyonjie replied:
It was from Kai, OK, changing it now.

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -395,6 +395,12 @@ struct snd_sof_dev {
+	 * and the suspending of d0i3_compatible stream during the S0ix
+	 * suspend/resume cycle, and clear the flag once resuming finished.
+	 */
+	bool s0_suspend;
```
@keyonjie what do you meanby "shared to different drivers" in the commit message. Let just simply say that the flag will be used to indicate that the system is entering S0ix. Even in the comment above, how and where it is used is not needed here.

plbossart replied:
Agree with @ranj063 the comment should be something like
'flag set in the "prepare" suspend stage to track the intended power target (S0 idle or S3)

keyonjie replied:
thank you both, changing it now.

ranj063 commented on sound/soc/sof/pm.c:
```diff
@@ -444,3 +444,27 @@ int snd_sof_suspend(struct device *dev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+
+	/* reset the s0_suspend flag */
```
comment is not needed. its quite self-explanatory

keyonjie replied:
OK.

ranj063 commented:
> My point was that since we don't support INFO_RESUME can this really happen in the first place?

@plbossart yeah it cant really happen in our case but when reading the PM documentation, it recommended that the driver do the needful and add the TRIGGER_RESUME as well for completeness I suppose.

plbossart commented:
> > My point was that since we don't support INFO_RESUME can this really happen in the first place?
> 
> @plbossart yeah it cant really happen in our case but when reading the PM documentation, it recommended that the driver do the needful and add the TRIGGER_RESUME as well for completeness I suppose.

yes, so let's mark it as being there for completeness but not a functional path.

keyonjie commented:
@plbossart @ranj063 Thanks for reviewing, I think I have addressed all comments, since the commits are out of order here, let me close this and create a new one.

keyonjie commented:
@plbossart @ranj063 now it is PR#1337
